cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java

```diff
@@ -207,31 +207,52 @@ private int getSystemServiceInstanceCount(String serviceName) throws Exception {
   }
 
   public void restartAllServiceInstances(HttpRequest request, HttpResponder responder, String serviceName) {
-    restartInstances(responder, serviceName, -1);
+    restartInstances(responder, serviceName, -1, true);
   }
 
   public void restartServiceInstance(HttpRequest request, HttpResponder responder, String serviceName,
                                      int instanceId) {
-    restartInstances(responder, serviceName, instanceId);
+    restartInstances(responder, serviceName, instanceId, false);
   }
 
-  private void restartInstances(HttpResponder responder, String serviceName, int instanceId) {
+  private void restartInstances(HttpResponder responder, String serviceName, int instanceId, boolean restartAll) {
     long startTimeMs = System.currentTimeMillis();
     boolean isSuccess = true;
+    if (!serviceManagementMap.containsKey(serviceName)) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Invalid service name %s", serviceName));
+      return;
+    }
+    MasterServiceManager masterServiceManager = serviceManagementMap.get(serviceName);
+
     try {
-      if (!serviceManagementMap.containsKey(serviceName)) {
-        responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Invalid service name %s", serviceName));
-        return;
+      if (!masterServiceManager.isServiceEnabled()) {
+        throw new IllegalStateException();
       }
 
-      MasterServiceManager masterServiceManager = serviceManagementMap.get(serviceName);
-      if (instanceId < 0) {
+      if (restartAll) {
         masterServiceManager.restartAllInstances();
       } else {
+        if (instanceId < 0 || instanceId >= masterServiceManager.getInstances()) {
+          throw new IllegalArgumentException();
```

Comment:
Instead, you can throw a ``BadRequestException``, and it will translate into 404. Check ``HttpExceptionHandler`` to see how these exceptions translate into responses.